client/service/src/client/client.rs

```diff
@@ -860,7 +862,11 @@ where
 							}
 						}
 
-						let state_root = operation.op.reset_storage(storage)?;
+						// This is use by fast sync so runtime version need to be resolve from
+						// changes.
+						let state_version =
+							Self::resolve_state_version_from_wasm(&storage, &self.executor)?;
+						let state_root = operation.op.reset_storage(storage, state_version)?;
```

Comment:
I'm not familiar with client code and don't understand this code, but couldn't there be some issue: let's say we do a runtime upgrade from state_version 0 to 1. In the block which does set_code we are still using state_version 0, but in storage we already have the new code to be used for next block. So when fetching state version from storage above we fetch from the runtime which wasn't to produce the block and which have a different state version.

Reply:
This part of code is specific to warp sync/fast sync (sc_consensus::StorageChanges::Import) where `storage` is the whole state we are injecting in block N.
So we want to use the block N-1 state, thus querying the payload instead of the state (reset_storage is replacing the whole state with N state).
So you are correct this won't work on a switch.
Yet it is not the only reason, if there was change of version, because we also do not support hybrid state, but this is a known limitation and one of the reason why applying migration is high priority after the switch.